<a href="https://colab.research.google.com/github/FernandoVina/Challenge_TelecomX/blob/main/ChallTelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [115]:
import pandas as pd
import requests
import json
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [197]:
# Carga de datos y DF
datos_telecom = pd.read_json('/content/TelecomX_Data.json')
datos_telecom.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [117]:
# Descripcion de tipo de datos
datos_telecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [118]:
# Revision de columnas de la base
datos_telecom.columns

Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [119]:
# Numero de filas y columnas
datos_telecom.shape

(7267, 6)

In [120]:
#revision del tipo de datos del objeto DF
type(datos_telecom)

pandas.core.frame.DataFrame

In [121]:
columnas = list(datos_telecom.columns)
columnas

['customerID', 'Churn', 'customer', 'phone', 'internet', 'account']

#🔧 Transformación

In [122]:
# se normaliza las columnas correspondientes
customer_data = pd.json_normalize(datos_telecom['customer'])
phone_data = pd.json_normalize(datos_telecom['phone'])
internet_data = pd.json_normalize(datos_telecom['internet'])
account_data = pd.json_normalize(datos_telecom['account'])

# Concatenando el nuevo dataframe con el original dataframe
datos_telecom_expanded = pd.concat([datos_telecom[['customerID', 'Churn']], customer_data, phone_data, internet_data, account_data], axis=1)

display(datos_telecom_expanded.head())

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [123]:
# Revision del tipo de datos de cada columna y su consistencia con sus datos
datos_telecom_expanded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [124]:
# cambio a numerico tipo float64, numero con decimales
 # errors='coerce' se aseguró de que cualquier valor en esa columna que no pudiera ser convertido a número
  # (como cadenas vacías o texto) fuera reemplazado por NaN (Not a Number), que representa valores faltantes.
datos_telecom_expanded['Charges.Total'] = pd.to_numeric(datos_telecom_expanded['Charges.Total'], errors='coerce')
print(datos_telecom_expanded.dtypes)

customerID           object
Churn                object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total       float64
dtype: object


In [125]:
datos_telecom_expanded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


# **Diccionario de datos**
* customerID: número de identificación único de cada cliente,
* Churn: si el cliente dejó o no la empresa,
* gender: género (masculino y femenino),
* SeniorCitizen: información sobre si un cliente tiene o no una edad igual o mayor a 65 años,
* Partner: si el cliente tiene o no una pareja,
* Dependents: si el cliente tiene o no dependientes,
* tenure: meses de contrato del cliente,
* PhoneService: suscripción al servicio telefónico,
*; MultipleLines: suscripción a más de una línea telefónica,
InternetService: suscripción a un proveedor de internet
* OnlineSecurity: suscripción adicional de seguridad en línea,
* OnlineBackup: suscripción adicional de respaldo en línea,
* DeviceProtection: suscripción adicional de protección del dispositivo,
* TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera,
* StreamingTV: suscripción de televisión por cable,
* StreamingMovies: suscripción de streaming de películas,
* Contract: tipo de contrato,
* PaperlessBilling: si el cliente prefiere recibir la factura en línea,
* PaymentMethod: forma de pago,
* Charges.Monthly: total de todos los servicios del cliente por mes,
* Charges.Total: total gastado por el cliente.

In [126]:
# Revision de algunas filas y todas las columnas, consistencia
datos_telecom_expanded.tail()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
7262               No      Mailed check           55.15         742.90  
7263              Yes  Electronic check           85.10        1873.70  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00        3707.60  

[5 rows x 21 columns]

In [127]:
#revisar duplicados
datos_duplizados = datos_telecom_expanded.duplicated().sum()
print(f"Número de filas duplicadas: {datos_duplizados}")

Número de filas duplicadas: 0


In [128]:
#verificar datos nulos
# datos_nulos = datos_telecom_expanded.isnull().values.sum()
datos_nulos = datos_telecom_expanded.isnull().sum()
print(f"Número de filas con valores nulos: {datos_nulos}")

Número de filas con valores nulos: customerID           0
Churn                0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
Charges.Monthly      0
Charges.Total       11
dtype: int64


In [129]:
# verificar si hay datos vacios
datos_vacios = datos_telecom_expanded.isna().sum()
print(f"Número de filas con valores vacíos: {datos_vacios}")

Número de filas con valores vacíos: customerID           0
Churn                0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
Charges.Monthly      0
Charges.Total       11
dtype: int64


In [130]:
# Verificacion si hay valores únicos en cada columna
for col in datos_telecom_expanded.columns:
  print(f"valores únicos en la columna'{col}': {datos_telecom_expanded[col].nunique()}")
  if datos_telecom_expanded[col].nunique() < 50: # Imprime los valores solamente
    print(datos_telecom_expanded[col].unique())
    print('-' * 50)  # Línea divisoria para mejor visualización

valores únicos en la columna'customerID': 7267
valores únicos en la columna'Churn': 3
['No' 'Yes' '']
--------------------------------------------------
valores únicos en la columna'gender': 2
['Female' 'Male']
--------------------------------------------------
valores únicos en la columna'SeniorCitizen': 2
[0 1]
--------------------------------------------------
valores únicos en la columna'Partner': 2
['Yes' 'No']
--------------------------------------------------
valores únicos en la columna'Dependents': 2
['Yes' 'No']
--------------------------------------------------
valores únicos en la columna'tenure': 73
valores únicos en la columna'PhoneService': 2
['Yes' 'No']
--------------------------------------------------
valores únicos en la columna'MultipleLines': 3
['No' 'Yes' 'No phone service']
--------------------------------------------------
valores únicos en la columna'InternetService': 3
['DSL' 'Fiber optic' 'No']
--------------------------------------------------
valores único

In [131]:
# verificacion de numero de filas y columnas
datos_telecom_expanded.shape

(7267, 21)

In [132]:
# Revisando datos nulos o NaN
datos_telecom_expanded.isnull().sum()

customerID           0
Churn                0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
Charges.Monthly      0
Charges.Total       11
dtype: int64

In [133]:
# la columna Charges.Total si tienen servicios y pago, pueden ser clienetes nuevos o temporales, se le incorporo como cero
# 0 dado que serian clientes nuevos, dado sus datos
null_charges_total = datos_telecom_expanded[datos_telecom_expanded['Charges.Total'].isnull()]
display(null_charges_total)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
975            No  No phone service             DSL  ...                  Yes   
1775          Yes                No              No  ...  No internet service   
1955          Yes               Yes             DSL  ...                  Yes   
2075          Yes                No              No  ...  No internet service   
2232          Yes                No              No  ...  No internet service   
2308          Yes               Yes              No  ...  No internet service   
2930          Yes               Yes             DSL  ...                  Yes   
3134          Yes               Yes              No  ...  No internet service   
3203           No  No phone service             DSL  ...                   No   
4169          Yes                No             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                   No                  Yes                   No   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                  Yes                  Yes                  Yes   
4169                  Yes                   No                  Yes   
5599  No internet service  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05            NaN  
1775               Mailed check           20.00            NaN  
1955  Bank transfer (automatic)           61.90            NaN  
2075               Mailed check           19.70            NaN  
2232               Mailed check           20.25            NaN  
2308               Mailed check           25.35            NaN  
2930               Mailed check           73.35            NaN  
3134               Mailed check           25.75            NaN  
3203  Ban

In [134]:
# Cambio los nulos en 'Charges.Total' por 0 y revision de cambio
datos_telecom_expanded['Charges.Total'] = datos_telecom_expanded['Charges.Total'].fillna(0)
display(datos_telecom_expanded.isnull().sum())

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

#📊 Carga y análisis

In [135]:
# Veamos calculos estadisticos
datos_telecom_expanded.describe()

SeniorCitizen       tenure  Charges.Monthly  Charges.Total
count    7267.000000  7267.000000      7267.000000    7267.000000
mean        0.162653    32.346498        64.720098    2277.182035
std         0.369074    24.571773        30.129572    2268.648587
min         0.000000     0.000000        18.250000       0.000000
25%         0.000000     9.000000        35.425000     396.200000
50%         0.000000    29.000000        70.300000    1389.200000
75%         0.000000    55.000000        89.875000    3778.525000
max         1.000000    72.000000       118.750000    8684.800000

In [136]:
# Contando los mayores y iguales a 65 años y los que no
senior_citizen_counts = datos_telecom_expanded['SeniorCitizen'].value_counts()
display(senior_citizen_counts)

SeniorCitizen
0    6085
1    1182
Name: count, dtype: int64

In [137]:
# Numero de clientes, no clientes y datos vacios
churn_counts = datos_telecom_expanded['Churn'].value_counts()
display(churn_counts)

Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

In [138]:
# datos vacios si tienen servicios en general, se registraran como No
churn_empty = datos_telecom_expanded[datos_telecom_expanded['Churn'] == '']
print(f"Number of empty 'Churn' values: {len(churn_empty)}")
display(churn_empty)

Number of empty 'Churn' values: 224


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30    0047-ZHDTW        Female              0      No         No      11   
75    0120-YZLQA          Male              0      No         No      71   
96    0154-QYHJU          Male              0      No         No      29   
98    0162-RZGMZ        Female              1      No         No       5   
175   0274-VVQOQ          Male              1     Yes         No      65   
...          ...   ...     ...            ...     ...        ...     ...   
7158  9840-GSRFX        Female              0      No         No      14   
7180  9872-RZQQB        Female              0     Yes         No      49   
7211  9920-GNDMB          Male              0      No         No       9   
7239  9955-RVWSC        Female              0     Yes        Yes      67   
7247  9966-VYRTZ        Female              0     Yes        Yes      31   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
30            Yes               Yes     Fiber optic  ...                   No   
75            Yes                No              No  ...  No internet service   
96            Yes                No             DSL  ...                  Yes   
98            Yes                No             DSL  ...                  Yes   
175           Yes               Yes     Fiber optic  ...                  Yes   
...           ...               ...             ...  ...                  ...   
7158          Yes               Yes             DSL  ...                  Yes   
7180           No  No phone service             DSL  ...                   No   
7211          Yes               Yes     Fiber optic  ...                   No   
7239          Yes                No              No  ...  No internet service   
7247          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
30                     No                   No                   No   
75    No internet service  No internet service  No internet service   
96                     No                  Yes                   No   
98                     No                  Yes                   No   
175                   Yes                   No                  Yes   
...                   ...                  ...                  ...   
7158                   No                   No                   No   
7180                   No                   No                  Yes   
7211                   No                   No                   No   
7239  No internet service  No internet service  No internet service   
7247  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
30    Bank transfer (automatic)           79.00         929.30  
75      Credit card (automatic)           19.90        1355.10  
96             Electronic check           58.75        1696.20  
98      Credit card (automatic)           59.90         287.85  
175   Bank transfer (automatic)          103.15        6792.45  
...                         ...             ...            ...  
7158               Mailed check           54.25         773.20  
71

In [139]:
# clientes con facturacion se incluyeron en los que No dejaron la empresa, son clientes con pagos y servicios vigentes
datos_telecom_expanded['Churn'] = datos_telecom_expanded['Churn'].replace('','No')
display(datos_telecom_expanded['Churn'].value_counts())

Churn
No     5398
Yes    1869
Name: count, dtype: int64

In [198]:
# Porcentaje de clientes retirados
(1869/7267)*100

25.7190037154259

In [141]:
# Verifiacion de clientes modificados que tieen cargos y servicios como 'No'
churn_no_zero_charges = datos_telecom_expanded[(datos_telecom_expanded['Churn'] == 'No') & ((datos_telecom_expanded['Charges.Monthly'] == 0)  | (datos_telecom_expanded['Charges.Total'] == 0))]
display(churn_no_zero_charges)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
975            No  No phone service             DSL  ...                  Yes   
1775          Yes                No              No  ...  No internet service   
1955          Yes               Yes             DSL  ...                  Yes   
2075          Yes                No              No  ...  No internet service   
2232          Yes                No              No  ...  No internet service   
2308          Yes               Yes              No  ...  No internet service   
2930          Yes               Yes             DSL  ...                  Yes   
3134          Yes               Yes              No  ...  No internet service   
3203           No  No phone service             DSL  ...                   No   
4169          Yes                No             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                   No                  Yes                   No   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                  Yes                  Yes                  Yes   
4169                  Yes                   No                  Yes   
5599  No internet service  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05            0.0  
1775               Mailed check           20.00            0.0  
1955  Bank transfer (automatic)           61.90            0.0  
2075               Mailed check           19.70            0.0  
2232               Mailed check           20.25            0.0  
2308               Mailed check           25.35            0.0  
2930               Mailed check           73.35            0.0  
3134               Mailed check           25.75            0.0  
3203  Ban

In [142]:
# Creacion de columna de valor diario a contar del valor mensual
datos_telecom_expanded['Charges.Daily'] = datos_telecom_expanded['Charges.Monthly'] / 30
display(datos_telecom_expanded.tail())

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... DeviceProtection  \
7262          Yes                No             DSL  ...               No   
7263          Yes               Yes     Fiber optic  ...               No   
7264          Yes                No             DSL  ...               No   
7265          Yes                No             DSL  ...              Yes   
7266           No  No phone service             DSL  ...              Yes   

     TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
7262         Yes          No              No        One year               No   
7263          No          No             Yes  Month-to-month              Yes   
7264          No          No              No  Month-to-month              Yes   
7265         Yes          No             Yes        Two year               No   
7266          No         Yes             Yes        Two year               No   

         PaymentMethod Charges.Monthly Charges.Total  Charges.Daily  
7262      Mailed check           55.15        742.90       1.838333  
7263  Electronic check           85.10       1873.70       2.836667  
7264      Mailed check           50.30         92.75       1.676667  
7265      Mailed check           67.85       4627.65       2.261667  
7266  Electronic check           59.00       3707.60       1.966667  

[5 rows x 22 columns]

In [143]:
# Verificacion de tipo de datos con nueva columna o campo cargo diario
datos_telecom_expanded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [144]:
# Crea df de retirados para analisis
datos_retirados = datos_telecom_expanded[datos_telecom_expanded['Churn'] == 'Yes']
display(datos_retirados.head())

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
2   0004-TLHLJ   Yes    Male              0      No         No       4   
3   0011-IGKFF   Yes    Male              1     Yes         No      13   
4   0013-EXCHZ   Yes  Female              1     Yes         No       3   
18  0022-TCJCI   Yes    Male              1      No         No      45   
19  0023-HGHWL   Yes    Male              1      No         No       1   

   PhoneService     MultipleLines InternetService  ... DeviceProtection  \
2           Yes                No     Fiber optic  ...              Yes   
3           Yes                No     Fiber optic  ...              Yes   
4           Yes                No     Fiber optic  ...               No   
18          Yes                No             DSL  ...              Yes   
19           No  No phone service             DSL  ...               No   

   TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
2           No          No              No  Month-to-month              Yes   
3           No         Yes             Yes  Month-to-month              Yes   
4          Yes         Yes              No  Month-to-month              Yes   
18          No          No             Yes        One year               No   
19          No          No              No  Month-to-month              Yes   

              PaymentMethod Charges.Monthly Charges.Total  Charges.Daily  
2          Electronic check            73.9        280.85       2.463333  
3          Electronic check            98.0       1237.85       3.266667  
4              Mailed check            83.9        267.40       2.796667  
18  Credit card (automatic)            62.7       2791.50       2.090000  
19         Electronic check            25.1         25.10       0.836667  

[5 rows x 22 columns]

In [145]:
# Cantuidad de retirados
print(datos_retirados.shape[0])

1869


In [203]:
# Veamos calculos estadisticos de Si retirados
datos_telecom_expanded[datos_telecom_expanded['Churn'] == 'Yes'].describe()

SeniorCitizen       tenure  Charges.Monthly  Charges.Total  \
count    1869.000000  1869.000000      1869.000000    1869.000000   
mean        0.254682    17.979133        74.441332    1531.796094   
std         0.435799    19.531123        24.666053    1890.822994   
min         0.000000     1.000000        18.850000      18.850000   
25%         0.000000     2.000000        56.150000     134.500000   
50%         0.000000    10.000000        79.650000     703.550000   
75%         1.000000    29.000000        94.200000    2331.300000   
max         1.000000    72.000000       118.350000    8684.800000   

       Charges.Daily  
count    1869.000000  
mean        2.481378  
std         0.822202  
min         0.628333  
25%         1.871667  
50%         2.655000  
75%         3.140000  
max         3.945000

In [147]:
# Veamos calculos estadisticos No retirados
datos_telecom_expanded[datos_telecom_expanded['Churn'] == 'No'].describe()

SeniorCitizen       tenure  Charges.Monthly  Charges.Total  \
count    5398.000000  5398.000000      5398.000000    5398.000000   
mean        0.130789    37.321045        61.354224    2535.263977   
std         0.337201    24.178367        31.105006    2330.802193   
min         0.000000     0.000000        18.250000       0.000000   
25%         0.000000    14.000000        25.100000     564.362500   
50%         0.000000    37.000000        64.500000    1663.625000   
75%         0.000000    61.000000        88.637500    4237.275000   
max         1.000000    72.000000       118.750000    8672.450000   

       Charges.Daily  
count    5398.000000  
mean        2.045141  
std         1.036834  
min         0.608333  
25%         0.836667  
50%         2.150000  
75%         2.954583  
max         3.958333

In [205]:
total_monthly_charges_all = datos_telecom_expanded['Charges.Monthly'].sum()
print(f"Total Monthly Charges de Total de Clientes: ${total_monthly_charges_all:.2f}")

Total Monthly Charges de Total de Clientes: $470320.95


In [206]:
total_monthly_charges_churned = datos_retirados['Charges.Monthly'].sum()
print(f"Total Monthly Charges de Clientes Retirados: ${total_monthly_charges_churned:.2f}")

Total Monthly Charges de Clientes Retirados: $139130.85


In [207]:
# Calculo de retirados % sobre tota l de Cargos Mensuales
(total_monthly_charges_churned / total_monthly_charges_all) * 100


np.float64(29.58210770751335)

In [150]:
#comparacion de la base por 'Churn' y su distribución
columns_to_compare = ['SeniorCitizen', 'gender', 'Partner', 'Dependents', 'PhoneService',
                        'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
                        'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
                        'Contract', 'PaperlessBilling', 'PaymentMethod']

for col in columns_to_compare:
    print(f"Comparing '{col}':")
    print("Full Dataset:")
    display(datos_telecom_expanded[col].value_counts())
    print("\nChurned Customers Dataset:")
    display(datos_retirados[col].value_counts())
    print("-" * 30)

Comparing 'SeniorCitizen':
Full Dataset:


SeniorCitizen
0    6085
1    1182
Name: count, dtype: int64


Churned Customers Dataset:


SeniorCitizen
0    1393
1     476
Name: count, dtype: int64

------------------------------
Comparing 'gender':
Full Dataset:


gender
Male      3675
Female    3592
Name: count, dtype: int64


Churned Customers Dataset:


gender
Female    939
Male      930
Name: count, dtype: int64

------------------------------
Comparing 'Partner':
Full Dataset:


Partner
No     3749
Yes    3518
Name: count, dtype: int64


Churned Customers Dataset:


Partner
No     1200
Yes     669
Name: count, dtype: int64

------------------------------
Comparing 'Dependents':
Full Dataset:


Dependents
No     5086
Yes    2181
Name: count, dtype: int64


Churned Customers Dataset:


Dependents
No     1543
Yes     326
Name: count, dtype: int64

------------------------------
Comparing 'PhoneService':
Full Dataset:


PhoneService
Yes    6560
No      707
Name: count, dtype: int64


Churned Customers Dataset:


PhoneService
Yes    1699
No      170
Name: count, dtype: int64

------------------------------
Comparing 'MultipleLines':
Full Dataset:


MultipleLines
No                  3495
Yes                 3065
No phone service     707
Name: count, dtype: int64


Churned Customers Dataset:


MultipleLines
Yes                 850
No                  849
No phone service    170
Name: count, dtype: int64

------------------------------
Comparing 'InternetService':
Full Dataset:


InternetService
Fiber optic    3198
DSL            2488
No             1581
Name: count, dtype: int64


Churned Customers Dataset:


InternetService
Fiber optic    1297
DSL             459
No              113
Name: count, dtype: int64

------------------------------
Comparing 'OnlineSecurity':
Full Dataset:


OnlineSecurity
No                     3608
Yes                    2078
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


OnlineSecurity
No                     1461
Yes                     295
No internet service     113
Name: count, dtype: int64

------------------------------
Comparing 'OnlineBackup':
Full Dataset:


OnlineBackup
No                     3182
Yes                    2504
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


OnlineBackup
No                     1233
Yes                     523
No internet service     113
Name: count, dtype: int64

------------------------------
Comparing 'DeviceProtection':
Full Dataset:


DeviceProtection
No                     3195
Yes                    2491
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


DeviceProtection
No                     1211
Yes                     545
No internet service     113
Name: count, dtype: int64

------------------------------
Comparing 'TechSupport':
Full Dataset:


TechSupport
No                     3582
Yes                    2104
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


TechSupport
No                     1446
Yes                     310
No internet service     113
Name: count, dtype: int64

------------------------------
Comparing 'StreamingTV':
Full Dataset:


StreamingTV
No                     2896
Yes                    2790
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


StreamingTV
No                     942
Yes                    814
No internet service    113
Name: count, dtype: int64

------------------------------
Comparing 'StreamingMovies':
Full Dataset:


StreamingMovies
No                     2870
Yes                    2816
No internet service    1581
Name: count, dtype: int64


Churned Customers Dataset:


StreamingMovies
No                     938
Yes                    818
No internet service    113
Name: count, dtype: int64

------------------------------
Comparing 'Contract':
Full Dataset:


Contract
Month-to-month    4005
Two year          1743
One year          1519
Name: count, dtype: int64


Churned Customers Dataset:


Contract
Month-to-month    1655
One year           166
Two year            48
Name: count, dtype: int64

------------------------------
Comparing 'PaperlessBilling':
Full Dataset:


PaperlessBilling
Yes    4311
No     2956
Name: count, dtype: int64


Churned Customers Dataset:


PaperlessBilling
Yes    1400
No      469
Name: count, dtype: int64

------------------------------
Comparing 'PaymentMethod':
Full Dataset:


PaymentMethod
Electronic check             2445
Mailed check                 1665
Bank transfer (automatic)    1589
Credit card (automatic)      1568
Name: count, dtype: int64


Churned Customers Dataset:


PaymentMethod
Electronic check             1071
Mailed check                  308
Bank transfer (automatic)     258
Credit card (automatic)       232
Name: count, dtype: int64

------------------------------


In [151]:
# Importo otro de graficos
import plotly.express as px

In [152]:
# Analizamos por sexo (retirados versus no retirados)
fig = px.histogram(datos_telecom_expanded, x='gender', text_auto= 'True', color = 'Churn', barmode = 'group', title='Distribución de Género en Clientes Retirados y No Retirados<br>Azul = no retirados, Rojo = retirados')
fig.update_layout(title_xanchor="center", title_x=0.5)
fig.show()

In [201]:
# Grafico de Distribución de Género en Clientes Iguales o Mayores a 65 años
fig = px.histogram(datos_telecom_expanded, x='SeniorCitizen', text_auto= 'True', color = 'Churn', barmode = 'group', title='Distribución de Género en Clientes Iguales o Mayores a 65 años<br>Azul = menores a 65 años, Rojo = Mayor o igual a 65 años')
fig.update_layout(title_xanchor="center", title_x=0.5)
fig.show()

In [202]:
# Grafico de Distribución de por Meses de Contrato
fig = px.histogram(datos_telecom_expanded, x='tenure', text_auto= 'True', color = 'Churn', barmode = 'group', title='Distribución de por Meses de Contrato<br>Azul = No Retirados, Rojo = Si Retirados')
fig.update_layout(title_xanchor="center", title_x=0.5)
fig.show()

In [181]:
# Revisión de Porcentajes
# Agrupar por tenure y Churn
df_porcentaje = datos_telecom_expanded.groupby(['tenure', 'Churn']).size().unstack(fill_value=0)


In [182]:
# Calcular el % de churn por tenure
df_porcentaje['churn_rate'] = (df_porcentaje.get('Yes', 0) / df_porcentaje.sum(axis = 1)) * 100

# Mostrar el resultado
print(df_porcentaje[['churn_rate']])

Churn   churn_rate
tenure            
0         0.000000
1        59.936909
2        50.000000
3        45.410628
4        44.864865
...            ...
68        8.653846
69        8.000000
70        9.016393
71        3.409091
72        1.626016

[73 rows x 1 columns]


In [165]:
# Graficamos los porcentajes
fig = px.line(df_porcentaje.reset_index(),
              x='tenure',
              y='churn_rate',
              markers = True,
              title='Porcentaje de Churn (Clientes Retirados) por Tenure (Meses de Contrato)')
fig.show()

In [183]:
# Que pasa si lo comparamos con el valor pago mensual
px.scatter(datos_telecom_expanded,
           x="tenure",
           y="Charges.Monthly",
           color="Churn",
           title="Tenure vs Cargos mensuales por Churn")

In [184]:
# Que pasa si lo comparamos con el valor pago mensual
px.scatter(datos_telecom_expanded,
           x="tenure",
           y="Charges.Daily",
           color="Churn",
           title="Tenure vs Cargos mensuales por Churn")

In [185]:
px.histogram(datos_telecom_expanded, x = 'Contract', text_auto = True, color = 'Churn', barmode = 'group')

In [187]:
px.box(datos_telecom_expanded, x = 'Churn', y = 'Charges.Monthly', color = 'Churn')

In [188]:
px.box(datos_telecom_expanded, x = 'Churn', y = 'Charges.Daily', color = 'Churn')

In [190]:
px.histogram(datos_telecom_expanded, x = 'PhoneService',  text_auto = True, color = 'Churn', barmode = 'group')

In [191]:
px.histogram(datos_telecom_expanded, x = 'InternetService',  text_auto = True, color = 'Churn', barmode = 'group')

In [193]:
px.histogram(datos_telecom_expanded, x = 'StreamingTV',  text_auto = True, color = 'Churn', barmode = 'group')

In [194]:
px.histogram(datos_telecom_expanded, x = 'PaymentMethod',  text_auto = True, color = 'Churn', barmode = 'group')

In [195]:
px.histogram(datos_telecom_expanded, x = 'Partner',  text_auto = True, color = 'Churn', barmode = 'group')

In [196]:
px.histogram(datos_telecom_expanded, x = 'Dependents',  text_auto = True, color = 'Churn', barmode = 'group')

#📄Informe final
Dada la revisión de los datos de la base de datos, se efectuaron varias modificaciones, a fin de obtener un mejor análsis de la misma. Los cambios estan comentados en cada código.
Dicho lo anterior podemos comentar lo siguiente:
* De la base total el 25% app. son retirados.
* El  30% del total de los ingresos mensuales corresponden a  los retirados.
* Existe un alto porcentaje de retiros en los primeros 4 meses, esto se deduce en el respectivo grafico.
* Tambien existe un alto numero de retiros en los clientes tipo Contract: tipo de contrato de Month-to-Month.
* Algo similar ocurre en los servicios Phone Service y Fiber Optic y
* Tambien con el tipo de pago Electronic Check.
* Estos ultimos tres puntos seria importante revisar su causa, una posible es facil de retirarse, por lo tanto importante aumentar el cruce de productos.
* Otro tema interesante de revisar la causa es que el los retiros con dependientes (30%)  es casi el doble de los no con dependiente (15%).